sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/BufferedExternalSorter.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.extensions.sorter;
+
+import java.io.IOException;
+import java.io.Serializable;
+import org.apache.beam.sdk.values.KV;
+
+/**
+ * Sorter that will use in memory sorting until the values can't fit into memory and will then fall
+ * back to external sorting.
+ */
+public class BufferedExternalSorter implements Sorter {
+  /** Contains configuration for the sorter. */
+  public static class Options implements Serializable {
+    private String tempLocation = "/tmp";
+    private int memoryMB = 100;
+
+    /** Sets the path to a temporary location where the sorter writes intermediate files. */
+    public void setTempLocation(String tempLocation) {
+      if (tempLocation.startsWith("gs://")) {
+        throw new IllegalArgumentException("Sorter doesn't support GCS temporary location.");
+      }
+
+      this.tempLocation = tempLocation;
+    }
+
+    /** Returns the configured temporary location. */
+    public String getTempLocation() {
+      return tempLocation;
+    }
+
+    /**
+     * Sets the size of the memory buffer in megabytes. This controls both the buffer for initial in
+     * memory sorting and the buffer used when external sorting.
+     */
+    public void setMemoryMB(int memoryMB) {
```

Comment:
any boundary conditions here? Presumably it needs to be >0. Should check this using `checkArgument`, and document it.